src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -11,21 +15,65 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
 
 
 public class ProductDetailsPageServlet extends HttpServlet {
     private ProductDao productDao;
 
+    private CartService cartService;
+
     @Override
     public void init(ServletConfig config) throws ServletException {
         super.init(config);
         productDao = ArrayListProductDao.getInstance();
+        cartService = DefaultCartService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String productId = request.getPathInfo();
-        request.setAttribute("product", productDao.getProduct(Long.valueOf(productId.substring(1))));
+        Long productId = parseProductId(request);
+        Product product = productDao.getProduct(productId);
+        List<Product> recentlyViewedProducts = (ArrayList<Product>) request.getSession().getAttribute("recentlyViewedProducts");
+        recentlyViewedProducts.remove(product);
+        recentlyViewedProducts.add(0, product);
+        if (recentlyViewedProducts.size() > 3) {
+            recentlyViewedProducts.remove(3);
+        }
+        request.setAttribute("cart", cartService.getCart(request));
+        request.setAttribute("product", productDao.getProduct(productId));
         request.getRequestDispatcher("/WEB-INF/pages/product.jsp").forward(request, response);
     }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        Long productId = parseProductId(request);
+        String quantityString = request.getParameter("quantity");
+        int quantity;
+        try {
+            NumberFormat format = NumberFormat.getInstance(request.getLocale());
+            quantity = format.parse(quantityString).intValue();
+        } catch (ParseException e) {
+            request.setAttribute("error", "Not a number");
+            doGet(request, response);
+            return;
+        }
+        Cart cart = cartService.getCart(request);
+        try {
```

Comment:
I'd combine this into single try with multiple catch. But it's just hint, no changes required